bodhi/server/consumers/pungi_wrapper.py

```diff
@@ -0,0 +1,348 @@
+            shutil.copyfileobj(variants_file_obj, fd)
+            variants_file_obj.close()
+
+    def write_repo_metadata(self):
+        # write treeinfo before ISOs are created
+        for variant in self.compose.get_variants():
+            for arch in variant.arches + ["src"]:
+                pungi.metadata.write_tree_info(self.compose, arch, variant)
+
+        # write .discinfo and media.repo before ISOs are created
+        for variant in self.compose.get_variants():
+            for arch in variant.arches + ["src"]:
+                timestamp = pungi.metadata.write_discinfo(self.compose, arch, variant)
+                pungi.metadata.write_media_repo(self.compose, arch, variant, timestamp)
+
+    def create_latest_repo_links(self):
+        latest_link = True
+
+        if latest_link:
+            self.compose_dir = os.path.basename(self.compose.topdir)
+            symlink_name = "latest-%s-%s" % (
+                self.compose.conf["release_short"],
+                self.compose.conf["release_version"]
+            )
+            if self.compose.conf["release_is_layered"]:
+                symlink_name += "-%s-%s" % (
+                    self.compose.conf["base_product_short"],
+                    self.compose.conf["base_product_version"]
+                )
+            symlink = os.path.join(self.compose.topdir, "..", symlink_name)
+
+            try:
+                os.unlink(symlink)
+            except OSError as ex:
+                if ex.errno != 2:
+                    raise Exception(ex)
+            try:
+                os.symlink(self.compose_dir, symlink)
+            except Exception as ex:
+                self.compose.log_error("Couldn't create latest symlink: %s" % ex)
+
+
+class VariantsConfig(object):
+
+    """
+    This class generates a variants config which can be used with the compose
+    object to generate a dnf repo. Right now it only works with modules.
+    """
+
+    def __init__(self, updates, builds, variant_id="Server",
+                 arches=['x86_64', 'armhfp', 'aarch64', 'i386', 'ppc64', 'ppc64le', 's390x']):
+        """ TODO: to be defined1. """
+        self.updates = updates
+        self.builds = builds
+        self.modules = self._generate_module_list()
+        self.arches = arches
+        self.headers = [
+            '<?xml version="1.0" encoding="UTF-8"?>',
+            ('<!DOCTYPE variants PUBLIC "-//Red Hat, Inc.//DTD '
+             'Variants info//EN" "variants2012.dtd">'),
+        ]
+        self.body = [
+            '<variants>',
+            '<variant id="%s" name="%s" type="variant">' % (variant_id, variant_id),
+            '<arches>',
+            ''.join(['<arch>%s</arch>' % arch for arch in self.arches]),
+            '</arches>',
+            '<modules>',
+            ''.join(['<module>%s</module>' % module for module in self.modules]),
+            '</modules>',
+            '</variant>',
+            '</variants>'
+        ]
```

Comment:
How much data will the modules be? This will all be held in memory if done this way.

Reply:
For modular release: https://pagure.io/pungi-fedora/blob/master/f/variants-modular.xml

Reply:
.. and we just don't know.  I hear ~20-40 modules for F27?  Of course it will grow over time.

Reply:
From video chat conversation - this seems like it will be fine.